text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+
+```rust
+use std::{env, io, sync::OnceCell};
+
+#[derive(Debug)]
+pub struct Logger { ... }
+
+static INSTANCE: OnceCell<Logger> = OnceCell::new();
+impl Logger {
+    pub fn global() -> &'static Logger {
+        INSTANCE.get().expect("logger is not initialized")
+    }
+    fn from_cli(args: env::Args) -> Result<Logger, std::io::Error> { ... }
+}
+
+fn main() {
+    let logger = Logger::from_cli(env::args()).unwrap();
+
+    // Note how we use locally-created value for initialization.
+    INSTANCE.set(logger).unwrap();
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies a common use-case.
+
+## Variations of `set`
+
+The RFC proposes "obvious" signature for the `set` method:
+
+```rust
+fn set(&self, value: T) -> Result<(), T>;
+```
+
+Note, however, that `set` establishes an invariant that the cell is initialized, so a more precise signature would be
+
+```rust
+fn set(&self, value: T) -> (&T, Option<T>);
+```
+
+To be able to return a reference, `set` might need to block a thread.
+For example, if two threads call `set` concurrently, one of them needs to block while the other moves the value into the cell.
+It is possible to provide a non-blocking alternative to `set`:
+
+```rust
+fn try_set(&self, value: T) -> Result<&T, (Option<&T>, T)>
+```
+
+That is, if value is set successfully, a reference is returned.
+Otherwise, ther the cell is either fully initialized, and a reference is returned as well, or the cell is being initialized, and no valid reference exist yet.
+
+## Support for `no_std`
+
+The RFC proposes to add `cell::OnceCell` and `cell::Lazy` to `core`, while keeping `sync::OnceCell` and `sync::Lazy` `std`-only.
+However, there's a subset of `sync::OnceCell` that can be provided in `core`:
+
+```rust
+impl<T> OnceCell<T> {
+    const fn new() -> OnceCell<T>;
+    fn get(&self) -> Option<&T>;
+    fn try_set(&self, value: T) -> Result<&T, (Option<&T>, T)>
+}
+```
+
+It is possible because, while `OnceCell` needs block for full API, its internal state can be implemented as a single `AtomicUsize`, so the `core` part does not need to know about blocking.
+It is unclear if this API would be significantly useful.
+In particular, the guarantees of non-blocking `set` are pretty weak, and are not enough to implement the `Lazy` wrapper.
```

Comment:
First this [post on reddit](https://old.reddit.com/r/rust/comments/cy910u/announcing_once_cell_100_single_assignment_cells/eyrfmna/) seems relevant:

> I am not an expert in the things that people use `#[no_std]` for, but my hypothesis is that, to make a robust `#[no_std]` library, one needs some more invasive changes than just swapping all mutexes for spin locks. I think that it is totally possible that a no_std target has some support for real blocking or preemption, and using spinning in this case seems wrong.
>
> A potential disaster movie scenario is this: there's no_std-only crate A which uses once_cell with default_features = false. An application B, which uses std and runs on linux, uses A. No other crate depends on once_cell, so its `std` feature is not activated, and it uses spinlock. A thread of B gets preempted while in `get_or_init` called from A. Another thread of B calls into A again, and is stuck busy-waiting. Granted, a lot of stars need to align to make this really bad, but the possibility still exists.
>
> On the other hand, if you are writing a no_std library, there's a high chance that you actually don't need blocking at all: in no_std your opportunities are limited, and whatever you do probably fits into lock free initialization. [rand/#50](https://github.com/rust-random/getrandom/issues/50) is a good example of this.
>
> If you are writing a no_std crate, and you are absolutely sure that you need blocking, I think it's best to parametrize your library by the `Mutex` trait, so that the users of your crate might plug it into whichever blocking facilities the host environment provides.

I wonder if spinning is always such a bad idea. `OnceCell::get_or_init` takes a closure that may take arbitrary long to execute, with the guarantee that is not running concurrently on another thread. But what if the initialization function is cheap and running it more than once doesn't matter? Then a thread could just start to run the closure first, and only spin for a few moments while the value is `memcpy`'d into place.

This would be pretty much the same as how a library would have to use `try_set` if it wants to support `no_std`, but knows as little about the target as `OnceCell`:
```rust
let cell = OnceCell::new();
let my_cheap_initialization = 21 + 21;
cell.try_set(my_cheap_initialization);
while cell.get().in_none() {}
```

This wouldn't combine well with `OnceCell`, because one thread could use our hypothetical initialization function, while another thread could do something else. But it is an option for `Lazy`, were the method of initialization is fixed at creation time:
```rust
impl<T, F> Lazy<T, F> {
    /// Creates a new lazy value with the given initializing function.
    pub const fn new(init: F) -> Lazy<T, F>;

    /// Creates a new lazy value with the given initializing function.
    /// The function may be executed by more than one thread at once, only the
    /// result of the first one to finish will be used.
    /// can be used in no_std.
    pub const fn new_concurrent(init: F) -> Lazy<T, F>;
}
```

This is just an idea, and a compromise. But as some `no_std` solution seems highly desired (see `lazy_static`), maybe something to consider.

Reply:
>But what if the initialization function is cheap and running it more than once doesn't matter? 

Yeah, this is an alternative design. If you are potentially ready to run initialization function several times and through away the result, you can even make this lock free, by boxing a value and atomically swapping the pointer. 

This is a valuable, but different, primitive. For std, I think it makes sense to lean towards more simple semantics, and guarantee that, while one thread runs initialization, all other block.

Reply:
> > But what if the initialization function is cheap and running it more than once doesn't matter?

> Yeah, this is an alternative design. If you are potentially ready to run initialization function several times and through away the result, you can even make this lock free, by boxing a value and atomically swapping the pointer.

> This is a valuable, but different, primitive. For std, I think it makes sense to lean towards more simple semantics, and guarantee that, while one thread runs initialization, all other block.

FWIW, this "valuable, but different" primitive would be ideal for *ring*'s needs for `Once`.